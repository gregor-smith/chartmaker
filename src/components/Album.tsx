import React, { FC } from 'react'
import { css, cx } from 'emotion'
import { Album as AlbumDetails, formatAlbumTitle } from '../state'


type Props = {
    details: AlbumDetails | null
    sizeRem: number
}


const baseStyle = css({
    backgroundColor: 'white',
    margin: '0.15rem'
})


export const Album: FC<Props> = ({ details, sizeRem }) => {
    const style = cx(
        baseStyle,
        css({
            width: `${sizeRem}rem`,
            height: `${sizeRem}rem`
        })
    )
    const image = details === null
        ? null
        : <img src={details.image.smallURL} alt={formatAlbumTitle(details)}/>
    return (
        <div className={style}>
            {image}
        </div>
    )
}
