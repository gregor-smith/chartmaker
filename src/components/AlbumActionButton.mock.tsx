const mock: typeof import('@/components/AlbumActionButton') = {
    AlbumActionButton: props =>
        <div {...props} className='mock-album-action-button'/>
}


export const AlbumActionButton = mock.AlbumActionButton
