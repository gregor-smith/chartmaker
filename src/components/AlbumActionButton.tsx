import { h, FunctionComponent } from 'preact'
import { css } from 'emotion'


type Props = {
    onClick: () => void
    title: string
}


const style = css({
    background: 'none',
    border: 'none'
})


export const AlbumActionButton: FunctionComponent<Props> = props =>
    <button {...props} class={style}/>
