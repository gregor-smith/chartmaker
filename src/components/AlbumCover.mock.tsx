import { cx } from 'emotion'


const mock: typeof import('@/components/AlbumCover') = {
    AlbumCover: ({
        album,
        size,
        overlayClass,
        children,
        highlighted,
        ...props
    }) => {
        const json = JSON.stringify(album)
        const className = cx('mock-album-cover', overlayClass)
        return (
            <div {...props} className={className}>
                {`Album: ${json}`}
                {`Size: ${size}`}
                {`Highlighted: ${highlighted}`}
                {children}
            </div>
        )
    }
}


export const AlbumCover = mock.AlbumCover
