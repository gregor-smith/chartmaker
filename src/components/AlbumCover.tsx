import React, { FC, DragEvent } from 'react'
import { css, cx } from 'emotion'

import { TEXT_COLOUR, ALBUM_PADDING_SIZE, highlightBackgroundStyle } from '@/style'
import { Album } from '@/types'


export type AlbumCoverProps = {
    album: Album
    size: string
    onDragStart: (event: DragEvent<HTMLDivElement>) => void
    onDragOver?: (event: DragEvent<HTMLDivElement>) => void
    onDragEnter?: (event: DragEvent<HTMLDivElement>) => void
    onDrop?: (event: DragEvent<HTMLDivElement>) => void
    onMouseEnter?: () => void
    overlayClass?: string
    highlighted?: boolean
}


const baseContainerStyle = css({
    position: 'relative',
    background: TEXT_COLOUR,
    margin: ALBUM_PADDING_SIZE
})

const imageStyle = css({
    position: 'absolute',
    top: 0,
    left: 0,
    width: '100%',
    maxWidth: '100%',
    height: '100%',
    objectFit: 'contain'
})

// the overlay is given an additional non-emotion class so it can be used as a
// selector in the container's style
const overlaySelector = 'overlay'

const baseOverlayStyle = cx(
    overlaySelector,
    imageStyle,
    css({ zIndex: 10 })
)


export const AlbumCover: FC<AlbumCoverProps> = ({
    album,
    size,
    onDragStart,
    onDragEnter,
    onDragOver,
    onDrop,
    onMouseEnter,
    children,
    overlayClass,
    highlighted
}) => {
    const overlayStyle = cx(baseOverlayStyle, overlayClass)
    const containerStyle = cx(
        baseContainerStyle,
        css({
            width: size,
            height: size,
            [`:not(:hover) .${overlaySelector}`]: {
                display: 'none'
            }
        }),
        highlighted === false
            ? highlightBackgroundStyle
            : undefined
    )

    let image: JSX.Element | undefined
    if (!album.placeholder) {
        image = (
            // lazy loading causes problems with html2canvas so keep it default
            <img className={imageStyle}
                src={album.url}
                alt={album.name}
                crossOrigin='anonymous'/>
        )
    }

    return (
        <div className={containerStyle}
                draggable={!album.placeholder}
                onDragStart={onDragStart}
                onDragEnter={onDragEnter}
                onDragOver={onDragOver}
                onDrop={onDrop}
                onMouseEnter={onMouseEnter}
                title={album.placeholder ? undefined : album.name}>
            {image}
            <div className={overlayStyle}>
                {children}
            </div>
        </div>
    )
}
