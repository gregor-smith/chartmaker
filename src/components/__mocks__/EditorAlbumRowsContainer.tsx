import type { FC } from 'react'

import type { EditorAlbumRowsContainerProps } from '@/components/EditorAlbumRowsContainer'


export const EditorAlbumRowsContainer: FC<EditorAlbumRowsContainerProps> = ({
    rows,
    highlighted
}) => {
    const json = JSON.stringify(rows)
    return (
        <div className='mock-editor-album-rows-container'>
            {`Rows: ${json}`}
            {`Highlighted: ${highlighted}`}
        </div>
    )
}
