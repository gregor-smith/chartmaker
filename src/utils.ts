import html2canvas from 'html2canvas'

import { BACKGROUND_COLOUR } from './style'


export function readFileText(file: File): Promise<string> {
    return new Promise((resolve, reject) => {
        const reader = new FileReader()
        reader.addEventListener('error', reject)
        reader.addEventListener('load', () => resolve(reader.result as string))
        reader.readAsText(file, 'utf-8')
    })
}


export function findIndex<T>(array: ReadonlyArray<T>, predicate: (item: T) => boolean): number | null {
    const index = array.findIndex(predicate)
    return index === -1 ? null : index
}


export async function elementToDataURI(element: HTMLElement, scale: number) {
    const canvas = await html2canvas(element, {
        allowTaint: true,
        scale,
        useCORS: true,
        imageTimeout: 0,
        backgroundColor: BACKGROUND_COLOUR
    })
    return canvas.toDataURL()
}


export function jsonToDataURI(json: string): string {
    return 'data:application/json;charset=utf-8,' + encodeURIComponent(json)
}


export function downloadURI(uri: string, filename: string) {
    const link = document.createElement('a')
    link.href = uri
    link.download = filename
    link.click()
    link.remove()
}
