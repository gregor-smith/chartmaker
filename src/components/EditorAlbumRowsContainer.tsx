import type { FC } from 'react'

import type { AlbumRow } from '@/state'
import type { Album } from '@/types'
import { EditorAlbumRow } from '@/components/EditorAlbumRow'
import type { DispatchProps } from '@/reducer'
import { AlbumRowsContainer } from '@/components/AlbumRowsContainer'


export type EditorAlbumRowsContainerProps = DispatchProps & {
    rows: AlbumRow<Album>[]
    highlighted: number | undefined
}


export const EditorAlbumRowsContainer: FC<EditorAlbumRowsContainerProps> = ({
    dispatch,
    rows,
    highlighted
}) => {
    const elements = rows.map(({ albums, size }, index) =>
        <EditorAlbumRow key={index}
            dispatch={dispatch}
            albums={albums}
            size={size}
            highlighted={highlighted}/>
    )
    return (
        <AlbumRowsContainer dispatch={dispatch}>
            {elements}
        </AlbumRowsContainer>
    )
}
