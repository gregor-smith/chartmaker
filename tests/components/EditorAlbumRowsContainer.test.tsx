import { render } from 'react-dom'

import type { AlbumRow } from '@/state'
import type { Album } from '@/types'
import { EditorAlbumRowsContainer } from '@/components/EditorAlbumRowsContainer'

import { createTestNamedAlbums, createTestPlaceholderAlbums, ignore, RenderContainer } from '../utils'


jest.mock('@/components/EditorAlbumRow')
jest.mock('@/components/AlbumRowsContainer')


const container = new RenderContainer()


test.each<[ AlbumRow<Album>[], number | undefined ]>([
    [
        [
            {
                albums: createTestNamedAlbums(3),
                size: '1rem'
            },
            {
                albums: createTestPlaceholderAlbums(5, 4),
                size: '2rem'
            }
        ],
        1
    ],
    [
        [
            {
                albums: createTestPlaceholderAlbums(3),
                size: '3rem'
            },
            {
                albums: createTestNamedAlbums(4, 4),
                size: '2rem'
            },
            {
                albums: createTestPlaceholderAlbums(2, 8),
                size: '1rem'
            }
        ],
        undefined
    ]
])('renders container and rows', (rows, highlighted) => {
    render(
        <EditorAlbumRowsContainer dispatch={ignore}
            rows={rows}
            highlighted={highlighted}/>,
        container.element
    )

    expect(container.element).toMatchSnapshot()
})
