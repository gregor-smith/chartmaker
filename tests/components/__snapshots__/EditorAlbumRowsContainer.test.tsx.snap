// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders container and rows 1`] = `
<div
  id="test-render-container"
>
  <div
    class="mock-album-rows-container"
  >
    <div
      class="mock-editor-album-title"
    >
      Albums: [{"id":1,"name":"Test album 1","url":"https://test.com/1"},{"id":2,"name":"Test album 2","url":"https://test.com/2"},{"id":3,"name":"Test album 3","url":"https://test.com/3"}]
      Size: 1rem
      Highlighted: 1
    </div>
    <div
      class="mock-editor-album-title"
    >
      Albums: [4,5,6,7,8]
      Size: 2rem
      Highlighted: 1
    </div>
  </div>
</div>
`;

exports[`renders container and rows 2`] = `
<div
  id="test-render-container"
>
  <div
    class="mock-album-rows-container"
  >
    <div
      class="mock-editor-album-title"
    >
      Albums: [1,2,3]
      Size: 3rem
      Highlighted: undefined
    </div>
    <div
      class="mock-editor-album-title"
    >
      Albums: [{"id":4,"name":"Test album 4","url":"https://test.com/4"}]
      Size: 2rem
      Highlighted: undefined
    </div>
    <div
      class="mock-editor-album-title"
    >
      Albums: [8,9]
      Size: 1rem
      Highlighted: undefined
    </div>
  </div>
</div>
`;
